### PR TITLE
Fix a bunch of issues in `To`

### PR DESCRIPTION
**Pull request by jto**

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala [hunk 1]
```diff
@@ -69,4 +70,17 @@ trait JavaInstances {
   implicit def javaBeanSchema[T: IsJavaBean: ClassTag]: RawRecord[T] =
     RawRecord[T](new JavaBeanSchema())
 
+  implicit def javaEnumSchema[T <: java.lang.Enum[T] : ClassTag]: Schema[T] =
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala]: TBH i think would prefer to use the straightforward way of using the `String` representation here. I see some problematic use case like when you are persist or have multiple pipelines running (i.e streaming) with "somehow" different Enum.

File: scio-core/src/main/scala/com/spotify/scio/schemas/SchemaMaterializer.scala [hunk 2]
```diff
@@ -115,7 +116,18 @@ object SchemaMaterializer {
     }
     record.construct(values)
   }
-  private def decode[A](schema: Type[A])(v: schema.Repr): A = v
+
+  private[scio] def decode[A](schema: Type[A])(v: schema.Repr): A = {
+    schema match {
+      case Type(t) if t.getTypeName() == BSchema.TypeName.LOGICAL_TYPE =>
+        // XXX: The Beam API is shitty
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/schemas/SchemaMaterializer.scala]: can we reword this? 😄

File: scio-core/src/main/scala/com/spotify/scio/schemas/To.scala [hunk 3]
```diff
@@ -35,13 +47,14 @@ object To {
       case (BSchema.TypeName.ARRAY, BSchema.TypeName.ARRAY, _) =>
         areCompatible(t0.getCollectionElementType, t1.getCollectionElementType)
       case (BSchema.TypeName.MAP, BSchema.TypeName.MAP, _) =>
-        areCompatible(t0.getMapKeyType, t1.getMapKeyType)
+        areCompatible(t0.getMapKeyType, t1.getMapKeyType) &&
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/schemas/To.scala]: whaa nice catch!

File: scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala (hunk 1 above)
Comment (sisidra) [reply scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala]: I agree, I think it also helps for use cases when writing beam sql query? I.e. user wants to write enum name, not ordinal. If I understand how this whole thing works :)

File: scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala [hunk 4]
```diff
@@ -69,4 +70,21 @@ trait JavaInstances {
...
+      private val className = clazz.getCanonicalName
+      private val values: Map[String, T] =
+        clazz.getMethod("values")
+          .invoke(null).asInstanceOf[Array[T]]
+          .map{ e => (e.name(), e) }
+          .toMap
+      override def getIdentifier: String = className
+      override def getBaseType: FieldType = FieldType.STRING
+      override def toBaseType(input: T): String = input.name()
+      override def toInputType(base: String): T = values(base)
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala]: awesome! I think we can even reduce this by just using:
 
```suggestion
      override def toInputType(base: String): T = java.lang.Enum.valueOf(clazz, base)
```

Comment (jto) [reply scio-core/src/main/scala/com/spotify/scio/schemas/instances/JavaInstances.scala]: Good catch. Apparently I can't Java 😅

File: scio-test/src/test/java/com/spotify/scio/sql/j/Level.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.spotify.scio.sql.j;
+
+public enum Level {
+  LOW, MEDIUM, HIGH
+}
```
Comment (regadas) [inline scio-test/src/test/java/com/spotify/scio/sql/j/Level.java]:
```suggestion
}

```

**Pull request merged**